pkg/kubelet/cm/devicemanager/topology.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package devicemanager
+
+import (
+	"reflect"
+
+	"k8s.io/klog"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/socketmask"
+
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1beta1"
+)
+
+type topology struct {
+	largestSocket int64
+	deviceMask    []socketmask.SocketMask
+	affinity      bool
+	allDevices    map[string][]pluginapi.Device
+}
+
+// GetTopologyHints calculates a BitMask based on device socket availability for an incoming container
+func (m *ManagerImpl) GetTopologyHints(pod v1.Pod, container v1.Container) topologymanager.TopologyHints {
```

Comment:
One example is, I do think this part of implementation could be decoupled from `topology.go` as specific impl for NUMA.

In detail, could we have sth like: `pkg/kubelet/cm/devicemanager/{tolology.go, numa.go, nv_link.go}`?

Reply:
@resouer would it be possible to get a diagram/description on how the nv link would work with device plugins & planned topology manager?

With NV Link, the idea is that two devices have an affinity is my understanding whereas the current proposal deals only with affinity between cpu and devices on a socket level. 

With an understanding of how NV Link could work with this proposal I'll have a better idea on how to abstract the current implementation to allow for such an extension in the future,

Reply:
@resouer  can you look at the updates to this PR? The socketmask implementation has changed and so the topology has changed